enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,458 @@
+The idea is to introduce two new features in LSO
+  - Auto discovery of local devices
+    - This will involve the intoduction of two new CRs `LocalVolumeDiscovery` and `LocalVolumeNodeDiscovery`.
+    - The purpose of this will be to expose local devices available in a node via the `LocalVolumeNodeDiscovery` CR.
+    - The device discovery will be done once via a discovery job.
+  - Auto discovery and provisioning of localDevices
+    - This will involve introduction of a new CR called `AutoDetectVolume`.
+    -  The pupose of this will be to auto discover and provision PVs on devices which match the inclusion filter present in the CR.
+    -  This will be a continous process.
+
+Once we have the detected devices the administrator can create the localPVs by explicitly selecting the disks. This can be done by the `localVolume` CR. The other option would be to create localPVs via the AutoDetectVolume CR and passing the inclusion filters in it.
+
+
+## Design Details for `Auto discovery of local devices`
+
+API scheme for `LocalVolumeDiscovery` and `LocalVolumeNodeDiscovery` CR:
+
+```go
+type LocalVolumeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscovery `json:"items"`
+}
+
+type LocalVolumeDiscovery struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoverySpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoverySpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type LocalVolumeDiscoveryStatus struct {
+	Phase string `json:"phase,omitempty"`
+	// RelatedObjects is a list of objects created and maintained by this
+	// operator. Object references will be added to this list after they have
+	// been created AND found in the cluster.
+	// +optional
+	RelatedObjects []corev1.ObjectReference `json:"relatedObjects,omitempty"`
+}
+
+type LocalVolumeNodeDiscoveryList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeNodeDiscovery `json:"items"`
+}
+
+type LocalVolumeNodeDiscovery struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              NodeDeviceMapSpec   `json:"spec"`
+	Status            NodeDeviceMapStatus `json:"status,omitempty"`
+}
+
+type NodeDeviceMapSpec struct {
+	NodeName string `json:"nodeName"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
```

Comment:
Why do we need tolerations field on this? The parent object is only designed to return devices discovered isn't it?

Reply:
The operator is going to start a Job on the requested node that will need tolerations for any taints, or else it won't run there. Should the operator automatically add all necessary tolerations or should we require the tolerations on the spec like this? Seems like we should do this automatically, otherwise the job will fail.

Reply:
User shouldn't have to specify tolerations on `LocalVolumeDiscovery` because that is the CR that will trigger device discovery. The toleration should simply be applied to oneshot job that is created on the node to handle discovery. We don't need to copy the toleration in CR that holds result of device discovery.

Reply:
+1 for not copying the toleration here.

Reply:
Regarding the tolerations... Since the intention of discovery is to run it once as a job, note that we won't be able to simply apply the node selector that was in the LocalVolumeDiscovery CR. Jobs don't allow us to run a single job on every node that matches the node selector. We will need to evaluate the node selector in order create Jobs that are have a node selector to tie each job to a node and also pass the tolerations on to each job.

Reply:
Thinking through.. I don't particularly like creating 1000 individual jobs and waiting for their completion. I think we can still use daemonset with a node selector and when number of created `LocalVolumeNodeDiscovery`(with proper status) matches number of replicas in daemonset we could just delete the daemonset.  It does require some more work in controller but it might be cleaner than spinning 1000 individual jobs.

But regardless of this - why can't we copy toleration specified in `LocalVolumeDiscovery` CR to `Job` object that will be created? Why do we have to copy toleration in `LocalVolumeNodeDiscovery` ?

Reply:
we were copying the tolerations to `LocalVolumeNodeDiscovery` because the controller of `LocalVolumeNodeDiscovery` was responsible for creating the jobs. The idea was user can provision the jobs by just creating the `LocalVolumeNodeDiscovery` CR for a node, But looks like the argument is that we should be using `LocalVolumeDiscovery` CR's controller for triggering the jobs, so removing it.

Reply:
So to reiterate what we discussed in this thread and on the Blue jeans calls over the last couple  of weeks: 

- Two main CRs used in the disk discovery:
a) **LocalVolumeDiscovery CR**: 
   - This is the main CR that will trigger the device discovery
   - It will create a DaemonSet. **We not going to create jobs on each node**. 
   - Toleration will be applied to this DaemonSet to ensure that the DaemonSet pod runs only on correct (selected) nodes.
   - The daemon set pod should identify the devices available on the node. And update the `Status` field of the LocalVolumeDiscoveryResult CR with these devices.
   - **Deletion of the DaemonSet:**
      - To find the count of DaemonSet pods running, we need to know the the number of valid nodes. 
      - Lets say we have 20 nodes that are running the DaemonSet pod
      - We should wait for 20 DaemonSet pods to have appropriate status (signifying that they ran successfully).
      - Now we can delete the DaemonSet.

b) **LocalVolumeDiscoveryResult CR ** :
  - We will be having one LocalVolumeDiscoveryResult CR per node. 
  - The `Status` field of this CR will be updated the appropriate device details that was discovered by LocalVolumeDiscovery CR above. 

@travisn @gnufied  could please confirm if this is what we are looking for?

Reply:
> We should wait for 20 DaemonSet pods to have appropriate status (signifying that they ran successfully).

The way we know if a pod is completed on a node is if there is a LocalVolumeDiscoveryResult on that node. I agree with your statement of what happens, but just to clarify, seems like we would need the following:
1. The LocalVolumeDiscovery CR is created
1. The operator remembers the current timestamp
1. The operator starts the daemonset 
1. The operator waits for some time, then queries if there is a LocalVolumeDiscoveryResult for each node where the discovery pods are running **and** that the timestamp on the status is newer than when the discovery was started.
   - If all results are not available, keep waiting and try again
1. When completed, delete the daemonset